### ISSUE TITLE
Job management

### ISSUE THREAD

**Issue by ctk3b** [mosdef-hub/msibi#12]
For now it seems things are more or less working if you assume that the user is running on a single node (but even then there are the issues of selecting multiple GPUs). I'm mostly opening this as a place for discussing on how to improve the situation.
- Has a similar problem already been solved? 
- Some kind of job management library could turn out to be very useful, e.g. http://ccl.cse.nd.edu/software/workqueue/
- What's the most portable way to detect GPUs on a given node?

**Comment by chrisiacovella** [comment]
On the queuing side: PBS is able to support allocating gpus; it will create a file PBS_GPUFILE which is similiar to PBS_NODEFILE. On degennes, I wrote some scripts to parse this file so you can get which gpu you should run on.  However, Raham doesn't use the first-in, first-out scheduler that comes with PBS, and instead uses Maui for scheduling, which sadly, breaks the PBS_GPUFILE aspect of PBS, but gives you other useful features, in particular, the ability to assign priority to queues (and implement fairshare for users). I talked with Will about SLURM, but the effort to get that working seems too much, and they didn't have much insight into dealing with the gpu allocation, since they just set their nodes to be job exclusive (that is, if you request a gpu node, you get all the gpus/cores).  All of the gpu-enabled systems I've worked on have all basically been job exclusive, which is what I think we will stick with. 

However, related to figuring out the gpus on a given node, you can always parse the output from nvidia-smi.  Although I've not tried to figure out how to do this on a multi-node run, since you'd need this info from all the different nodes you are asking for.  

Since MSIBI is effectively spawning jobs, I agree some sort of job management code would be useful to keep track of them.   I'm not sure the best way to do this, since you'd obviously want this code to be rather self-contained (i.e., not need to be installed on the cluster itself, but as part of the code).

**Comment by ctk3b** [comment]
> you can always parse the output from nvidia-smi
@tcmoore3, didn't you find a case where this broke? If not then I guess that problem is [already solved](https://github.com/ctk3b/msibi/blob/master/msibi/workers.py#L77) for nvidia systems.

**Comment by chrisiacovella** [comment]
They intentionally broke some of the nicer features for the consumer cards, so you can't see which processors are using which gpu.  However, you can still figure this out which gpu is being tasked via the memory usage. 

Check out:
https://developer.nvidia.com/nvidia-management-library-nvml

**Comment by ctk3b** [comment]
So we should probably be using [this](https://pypi.python.org/pypi/nvidia-ml-py/7.346.0) if it turns out to be stable instead of parsing the output of nvidia-smi. Hopefully it's also actively maintained...

**Comment by chrisiacovella** [comment]
That software looks reasonable.  Presumably we could write a simple piece of C/C++ code to access the library functions we specifically need as well, if it turns out not to be maintained.

**Comment by tcmoore3** [comment]
`nvidia-smi` has been working on accre. Right now our jobs aren't running in parallel in the most sophisticated manner (e.g., all jobs in a "chunk" finish before the next chunk of jobs is launched). That's something we need to figure out on the MS IBI side. As far as allocating resources, this nvidia-ml-py seems to do what we want, and is somewhat actively maintained.

On a positive note, I got MSIBI up and running on rahman. On a less positive note, it's not properly allocating GPUs. It's doing the different simulations in parallel, 2 at a time (2 GPUs), but it's assigning the same GPU to each job. Clearly, something is amiss, but I haven't yet taken the time to thoroughly investigate.

**Comment by chrisiacovella** [comment]
When we were running on keeneland, there were 3 gpus per node.  These were nanowire elongation runs that would finish whenever the wire broke (so each job walltime varied a lot).  What I did was have my script break up the jobs into 3 subscripts (one per gpu).  So gpu=0 got a list of jobs to run, gpu=1, got a list, etc.  Each list was backgrounded and thus didn't depend on the others (until they were all finished and encountered the wait statement, since I was using bash).   We were running ~100 jobs, so this approach worked, I'm assuming MSIBI will have <10 that need to run concurrently, so that might not be the best plan.

**Comment by chrisiacovella** [comment]
Clearly, what we need is something like PBS but for each gpu that the code can call. But for portability, it needs to be stand alone software because most people can't install stuff on clusters. 

I'm sure we can do this if we have a shared datastructure that keeps track of which jobs are running, on what gpu, and which are done, all embedded in a parallel while loop.  This should be relatively easy with openmp: asynchronous thread execution (including the ability to lock the shared array if you are grabbing a job from it or changing the status of a job), with only one blocking statement after the entire loop has been completed.

I'm going to mess around with writing a simple code (that doesn't actually really run anything, just does the threading).

**Comment by chrisiacovella** [comment]
It looks like nvidia-smi is working properly on Rahman.  I basically plucked out the essential bits of code from the worker script and tested them.  It was telling me two gpus, and assigning me the right one based on idx in the loop.  

I think someone needs double check what idx is being assigned in the actual worker routine and also set up a minimal test of the multiprocessing part (make sure there isn't a bug in the python version we are using).

**Comment by chrisiacovella** [comment]
So the issue seems to be related to setting "chunk_size" in the imap function.  When you define the pool, you are setting the number of threads that can be run at a time.  When you define chunk_size, you are telling it how to partition the total number of tasks, as presumable you can get some speed increase by smart partitioning (for large number of jobs).  However, since idx, which defines which card to use, is not defined based on what chunk you end up in, you end up having cases where it's trying to run all jobs that were given an even idx (and hence running on the same card).  

Basically, if you sent chunk_size = 1, it solves the problem, since it doesn't try to group jobs, and it will just do them sequentially (i.e., the same manner idx is assigned).

**Comment by ctk3b** [comment]
Ok so let's keep the general discussion on ideas and resolve specifics in their own issues.

* Fixes for intra-node GPU selection: #20
* Job management across multiple nodes will require some more brainstorming.
* Once we start supporting more engines, we may have to talk about divvying up CPU simulations as well.

**Comment by chrisiacovella** [comment]
Well, more on the general topic of the multiprocessing library...can we spawn processes asynchronously?  Like if we tell it to run 10 total threads, 2 threads at a time, is the blocking operation at the end of executing all 10 threads, or after executing 2? 

Furthermore, can we control how it breaks up the threads into chunks?  We can obviously calculate the time to complete each process...these times are relatively consistent during the iteration.  We could group jobs by the approximate wall time they take, in order to load balance better.   By grouping it could be: (1) which two jobs run at the same time or (2) which jobs go into which chunk.   The gpu id to use could be dynamically set (have a shared memory array that says which gpu is idle/used), rather than being assigned initially. 

In terms of running on multiple nodes, gnu parallel might do the trick, since we are basically talking about a bunch of "serial" jobs.  This is also pretty lightweight code someone could build in their home account if it's not available on a cluster. This is nice because it just uses the pbs_nodefile, so there is nothing really tricky going on. We would probably have to assume a homogeneous environment (i.e., each node has the same number of gpus), since the current code queries nvidia_smi. 

https://wiki.scinet.utoronto.ca/wiki/index.php/User_Serial#GNU_Parallel

**Comment by ctk3b** [comment]
@tcmoore3 started messing with the asynchronous threading. Here's actually a good explanation that might help clear up some of the confusion we had: http://stackoverflow.com/a/26521507/2597468

However, let's say we get the workers submitted asynchronously, we're still managing the `gpu_id` manually which means, e.g., on a 2 GPU node job 1 and job 2 start running, 2 finishes and 3 gets launched but because of how we're defining `gpu_id`, it will try to run on top of job 1. To solve that we would either need to do some kind of approximation on the wall time like you suggested or otherwise we'd have to get the GPU activity somehow? 

GNU parallel looks pretty nice. Sounds like it'd be worth playing with on Rahman.

**Comment by chrisiacovella** [comment]
In the current code, the gpu is determined by doing the modulus of your job id and the number of available cards.   As such, if things are asynchronous, a job may run on an occupied gpu. 

I think the simple solution is to change the gpu assignment to make it dynamically assigned by which one isn't being used both other threads. We need a shared memory array, where each index represents a gpu.  If the value of the array = 0, the gpu corresponding to that index available, if it's 1, it's not available.  The thread just picks out which gpu is available and increments the array to 1 (and when it's done, it sets it to 0).  Obviously, this requires a thread to lock the array before reading (i.e., only one thread can access the array at a time, both for read and write operations)...making sure that is correct is probably the tricky part (again, I know how to do that in openmp, need to make a test script in python to get the logic correct).

**Comment by ctk3b** [comment]
That's actually much simpler. I think you can just use a normal dictionary since this part of the code is just threaded:
http://stackoverflow.com/questions/1312331/using-a-global-dictionary-with-threads-in-python

It may require using the lock like in the example shown but I don't think it will be particularly complicated.

**Comment by ctk3b** [comment]
Or maybe a queue is simpler? http://stackoverflow.com/a/1312704/2597468
